### ISSUE TITLE
Installation fails using Poetry because of dependency mismatch.

### ISSUE THREAD

**Issue by NomeChomsky** [piccolo-orm/piccolo#22]
```
Because no versions of piccolo-admin match >0.10.3,<0.11.0
   and piccolo-admin (0.10.3) depends on uvicorn (>=0.11.0,<0.12.0), piccolo-admin (>=0.10.3,<0.11.0) requires uvicorn (>=0.11.0,<0.12.0).
  So, because pico depends on both uvicorn (^0.13.0) and piccolo-admin (^0.10.3), version solving failed.
```

Poetry is really annoying that way - but otherwise is a great tool. Would be great to get this fixed.

**Comment by dantownsend** [comment]
@NomeChomsky Thanks for the heads up. I don't use Poetry personally, so haven't encountered this problem, but I know it's a popular tool. I'll fix it by loosening the dependencies a bit.

**Comment by dantownsend** [comment]
@NomeChomsky Try version 0.10.4 of Piccolo Admin - I've loosened the dependencies. Hopefully that solves the problem.

**Comment by NomeChomsky** [comment]
Fantastic dan.

So far so good with Piccollo. I've been looking for a tool like this - tightly coupled on the back end, with migrations, and loosely coupled on the front.

Piccollo looks promising, and I'm going to see how far I get with a project using it. I must say, I'm an SQL novice. I researched lots of python frameworks which are tightly coupled on the back end. Piccollo looks great. One feature I'd love to see is to integrate a 'human readable' style relationships tooling inspired by Emmett. His docs are really excellent, but so far I wasn't able to make it work seamlessly.

If you look at the way Emmett deals with relationships on models, it's a really nice implementation. I'd love to see this come to Piccollo, because defining relationships is sometimes very abstract and headachey.

For example, how might I deal with circular dependencies when passing the model into the foreignkey field? Can I pass in a string representation?

**Comment by NomeChomsky** [comment]
To add - I think the command integration is also a very nice idea, and it works nicely. The idea that there's 'a correct way' to integrate table modelling into the project, and that Piccollo will bring that template in is very nice. It's a nice mix between Django 'this is how its done' styling, and Flask's micro framework style. It strikes me as a nice balance.

The Admin tool is the icing on the cake,

**Comment by dantownsend** [comment]
@NomeChomsky I haven't used Emmett before, but just had a look at the docs. With the foreign key relationships in Piccolo, the ``ForeignKey`` column references another ``Table`` subclass.

```python
class Manager(Table):
    name = Varchar(length=50)


class Band(Table):
    name = Varchar(length=50)
    manager = ForeignKey(Manager, null=True)
```

If you need to have tables which reference themselves, you can do it like this:

```python
class Manager(Table):
    name = Varchar(length=50)
    manager = ForeignKey('self', null=True)
```

I see your point though about using string representations, because with very complex schemas you may get circular dependencies. This isn't something Piccolo currently supports, but it wouldn't be too hard to implement. I'll add it to the roadmap.

I'd like to create a tool which shows a graphic representation of the the Piccolo schema - possibly a web app, but that's in the future.

Thanks for your comments and interest in the project. I'm interesting in hearing any other thoughts you have as you build stuff with Piccolo.

**Comment by dantownsend** [comment]
FYI, there's a fix for this in Piccolo 0.14.10 - ``ForeignKey`` can now accept strings, which should prevent circular dependency issues.

```python
class Band(Table):
    name = Varchar()
    manager = ForeignKey('Manager')

class Manager(Table):
    name = Varchar()

```

https://piccolo-orm.readthedocs.io/en/latest/piccolo/schema/column_types.html#foreignkey

**Comment by hipertracker** [comment]
I found the auto migrations don't work when a model contains ForeignKey. I receive a cryptic: unhashable type: 'list' error inestead.
